driver-scala/src/main/scala/org/mongodb/scala/package.scala

```diff
@@ -392,9 +399,18 @@ package object scala extends ClientSessionImplicits with ObservableImplicits wit
 
   implicit def bsonDocumentToDocument(doc: BsonDocument): Document = new Document(doc)
 
-  implicit def bsonDocumentToUntypedDocument(doc: BsonDocument): org.bson.Document =
-    org.bson.Document.parse(doc.toJson())
+  implicit def documentToUntypedDocument(doc: Document): org.bson.Document =
+    bsonDocumentToUntypedDocument(doc.underlying)
+
+  implicit def bsonDocumentToUntypedDocument(doc: BsonDocument): org.bson.Document = {
+    val writer: BsonBinaryWriter = new BsonBinaryWriter(new BasicOutputBuffer())
+    new BsonDocumentCodec().encode(writer, doc, EncoderContext.builder().build())
```

Comment:
I am not sure, but at first glance`BsonDocumentCodec` looks like an immutable class. If that is the case, do we actually want to create a new object per the `bsonDocumentToUntypedDocument` method invocation? Unless we are fortunate and [scalar replacement](https://shipilev.net/jvm/anatomy-quarks/18-scalar-replacement/) kicks in, this creates unnecessary garbage.

Reply:
Good call - done.